iothub/device/src/Transport/Mqtt/OrderedTwoPhaseWorkQueue.cs

```diff
@@ -15,71 +16,79 @@ private class IncompleteWorkItem
+        public OrderedTwoPhaseWorkQueue(
+            Func<IChannelHandlerContext, TWork, Task> worker,
+            Func<TWork, TWorkId> getWorkId,
+            Func<IChannelHandlerContext, TWork, Task> completeWork)
             : base(worker)
         {
-            this.getWorkId = getWorkId;
-            this.completeWork = completeWork;
+            _getWorkId = getWorkId ?? throw new ArgumentNullException(nameof(getWorkId));
+            _completeWork = completeWork ?? throw new ArgumentNullException(nameof(completeWork));
         }
 
         public Task CompleteWorkAsync(IChannelHandlerContext context, TWorkId workId)
         {
-            if (this.incompleteQueue.Count == 0)
+            if (!_incompleteQueue.Any())
             {
                 throw new IotHubException("Nothing to complete.", isTransient: false);
             }
-            IncompleteWorkItem incompleteWorkItem = this.incompleteQueue.Peek();
-            if (incompleteWorkItem.Id.Equals(workId))
+
+            IncompleteWorkItem incompleteWorkItem = _incompleteQueue.Peek();
+            if (incompleteWorkItem != null
+                && incompleteWorkItem.Id.Equals(workId))
             {
-                this.incompleteQueue.Dequeue();
-                return this.completeWork(context, incompleteWorkItem.WorkItem);
+                _incompleteQueue.Dequeue();
+                return _completeWork(context, incompleteWorkItem.WorkItem);
             }
+
             throw new IotHubException(
                 $"Work must be complete in the same order as it was started. Expected work id: '{incompleteWorkItem.Id}', actual work id: '{workId}'",
                 isTransient: false);
         }
 
         protected override async Task DoWorkAsync(IChannelHandlerContext context, TWork work)
         {
-            this.incompleteQueue.Enqueue(new IncompleteWorkItem(this.getWorkId(work), work));
+            _incompleteQueue.Enqueue(new IncompleteWorkItem(_getWorkId(work), work));
             await base.DoWorkAsync(context, work).ConfigureAwait(false);
         }
 
         public override void Abort()
         {
-            this.Abort(null);
+            Abort(null);
         }
 
         public override void Abort(Exception exception)
         {
-            States stateBefore = this.State;
+            States stateBefore = State;
             base.Abort(exception);
-            if (stateBefore != this.State && this.State == States.Aborted)
+
+            if (stateBefore != State
+                && State == States.Aborted)
             {
-                Queue<IncompleteWorkItem> queue = this.incompleteQueue;
-                while (queue.Count > 0)
+                while (_incompleteQueue.Any())
                 {
-                    TWork workItem = queue.Dequeue().WorkItem;
+                    var workItem = _incompleteQueue.Dequeue()?.WorkItem as ICancellable;
```

Comment:
If the event more than one thread was aborting (despite the protection supposedly offered in the lines above), even if we thought there was work in the queue and then tried to dequeue it, we might as well protect against an empty queue.